File: richos/mobile/native-android/core/src/main/kotlin/dev/richos/android/core/Action.kt
package dev.richos.android.core

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Everything that can change [AppState]. The UI, the headless CLI and the emulator bridge
 * all produce these and all go through [RichCore.dispatch]; there is no second path.
 *
 * The JSON form is the preserved phone core's (`richos/mobile/core/app.js`), discriminated
 * by `type`: `{"type":"compose","text":"Hello"}`, `{"type":"send"}`, … so a scenario written
 * for one client reads the same on every client. `theme` is the one addition.
 */
@Serializable
sealed interface Action {
    @Serializable @SerialName("play-kept")
    data object PlayKept : Action
    @Serializable @SerialName("playback-ended")
    data class PlaybackEnded(val id: String) : Action
    @Serializable @SerialName("select-thread")
    data class SelectThread(val threadId: String) : Action

    @Serializable @SerialName("compose")
    data class Compose(val text: String) : Action

    @Serializable @SerialName("remember-reading")
    data class RememberReading(val anchor: ReadingAnchor? = null) : Action

    @Serializable @SerialName("send")
    data object Send : Action

    @Serializable @SerialName("send-voice")
    data class SendVoice(val recording: Recording, val clientId: String? = null, val threadId: String? = null) : Action

    @Serializable @SerialName("network")
    data class Network(val online: Boolean) : Action

    @Serializable @SerialName("retry")
    data object Retry : Action

    @Serializable @SerialName("sync")
    data object Sync : Action

    @Serializable @SerialName("discard")
    data class Discard(val clientId: String) : Action

    @Serializable @SerialName("theme")
    data class SetTheme(val theme: Theme) : Action

    /** A scanned or pasted pairing link (contract §2.1): sends the code to the Mac it names. */
    @Serializable @SerialName("pair")
    data class Pair(val link: String) : Action

    /** The user's answer to the six words: "They match" (true) or "They do not match" (false). */
    @Serializable @SerialName("confirm-words")
    data class ConfirmWords(val match: Boolean) : Action

    /**
     * Pairing v2: ask the Mac whether the person has pressed "They match" on it, if an ask is due
     * (the app's one timer sends this at [AppState.macWaitDueInMs]). The core decides: nothing is
     * asked while the app is hidden, before the schedule, past the bound or past 22 requests.
     */
    @Serializable @SerialName("mac-wait")
    data object MacWait : Action

    /** Forget this pairing on the phone. Refused while unsent work is waiting. */
    @Serializable @SerialName("forget")
    data object Forget : Action

    /**
     * Bytes from the Mac's event stream, exactly as they arrived, split anywhere (contract §5.4).
     * The connection owner feeds the live stream through this; a scenario or the CLI can feed a
     * recorded one, which is how the conversation is tested without a Mac.
     */
    @Serializable @SerialName("receive")
    data class Receive(val wire: String) : Action

    /** The stream owner's socket changed state: `opening`, `open` or `away`. */
    @Serializable @SerialName("link")
    data class Link(val status: LinkStatus) : Action

    /**
     * What a diagnosis found while away: the phone's own network, and the managed service. And
     * [vpn], the OS's own report of whether the default network runs through a VPN (Tailscale on
     * Android is one), from its network callback, never a probe (D05).
     */
    @Serializable @SerialName("health")
    data class Health(val phoneOnline: Boolean? = null, val service: ServiceState? = null, val vpn: Boolean? = null) : Action

    /** Time passed: republish time-derived state (the reconnecting notice, the voice timer). The app's one timer sends it. */
    @Serializable @SerialName("tick")
    data object Tick : Action

    // --- voice (round-12 groups 4, 5, 6). Names and fields are the native iOS core's. ----------

    /** Touch-down on the microphone. [id] names the recording; [width] is the composer's, in dp. */
    @Serializable @SerialName("voice-press")
    data class VoicePress(val id: String, val width: Double, val at: Long) : Action

    /** Accessibility's "record hands-free": start a recording already locked. */
    @Serializable @SerialName("voice-start-locked")
    data class VoiceStartLocked(val id: String, val width: Double, val at: Long) : Action

    /**
     * The OS answered or reports the microphone permission. [canAsk]: while denied, whether the
     * system would still show its own question if asked again (Android's
     * `shouldShowRequestPermissionRationale`); false once it will not, and always on iOS.
     */
    @Serializable @SerialName("microphone-permission")
    data class MicrophonePermission(
        val permission: Microphone,
        @OptIn(kotlinx.serialization.ExperimentalSerializationApi::class)
        @kotlinx.serialization.EncodeDefault(kotlinx.serialization.EncodeDefault.Mode.NEVER)
        val canAsk: Boolean = false,
    ) : Action

    /** The microphone-off card's "Allow microphone": ask the system again, only while it still would. */
    @Serializable @SerialName("ask-microphone")
    data object AskMicrophone : Action

    /** The microphone-off card's "Not now": the card goes until the next press finds the microphone off. */
    @Serializable @SerialName("dismiss-microphone-card")
    data object DismissMicrophoneCard : Action

    /** The finger moved: offsets from the touch-down point, in dp (negative = left / up). */
    @Serializable @SerialName("voice-move")
    data class VoiceMove(val dx: Double, val dy: Double, val at: Long) : Action

    @Serializable @SerialName("voice-release")
    data class VoiceRelease(val at: Long) : Action

    /** Tap on the send circle while locked. */
    @Serializable @SerialName("voice-locked-send")
    data class VoiceLockedSend(val at: Long) : Action

    /** Tap on Cancel while locked. */
    @Serializable @SerialName("voice-locked-cancel")
    data class VoiceLockedCancel(val at: Long) : Action

    /** The system took the touch away (an alert over the app). */
    @Serializable @SerialName("voice-touch-canceled")
    data class VoiceTouchCanceled(val at: Long) : Action

    /** The app left the screen or the OS took the audio: the recording is kept, never sent. */
    @Serializable @SerialName("voice-interrupted")
    data class VoiceInterrupted(val at: Long) : Action

    /** The recording's level, 0 to 1, sampled every 100 ms. */
    @Serializable @SerialName("voice-level")
    data class VoiceLevel(val level: Double) : Action

    /** The end animation finished. */
    @Serializable @SerialName("voice-settled")
    data object VoiceSettled : Action

    /** Send a kept recording (`rec-card`). */
    @Serializable @SerialName("send-kept")
    data class SendKept(val id: String, val at: Long = 0) : Action

    /** Let a kept recording go. */
    @Serializable @SerialName("discard-kept")
    data class DiscardKept(val id: String) : Action

    // --- notifications, settings, update notices (round-12 groups 8, 9, 10; iOS names) -----------

    @Serializable @SerialName("turn-on-notifications")
    data object TurnOnNotifications : Action

    /** What the OS and the Mac said to a registration (or its absence). */
    @Serializable @SerialName("notifications-result")
    data class NotificationsResult(val status: NotificationStatus) : Action

    @Serializable @SerialName("turn-off-notifications")
    data object TurnOffNotifications : Action

    /** `notif-offer`'s "Not now": asked once. */
    @Serializable @SerialName("dismiss-notification-offer")
    data object DismissNotificationOffer : Action

    @Serializable @SerialName("set-previews")
    data class SetPreviews(val on: Boolean) : Action

    @Serializable @SerialName("open-sheet")
    data class OpenSheet(val sheet: Sheet) : Action

    @Serializable @SerialName("close-sheet")
    data object CloseSheet : Action

    /** Settings' red row: opens `settings-forget`, or `settings-forget-blocked` while unsent work waits. */
    @Serializable @SerialName("forget-pairing")
    data object ForgetPairing : Action

    /** The forget sheet's confirmation: notifications off first, then the key and the pairing go. */
    @Serializable @SerialName("confirm-forget")
    data object ConfirmForget : Action

    @Serializable @SerialName("open-system-settings")
    data object OpenSystemSettings : Action

    /** A notification was tapped: open that conversation and let that reply glow once (`conv-focused`). */
    @Serializable @SerialName("opened-from-notification")
    data class OpenedFromNotification(val messageId: String, val threadId: String? = null) : Action

    @Serializable @SerialName("clear-focus")
    data object ClearFocus : Action

    /** The hosted policy's answer: a notice (or none) and whether voice is paused by policy. */
    @Serializable @SerialName("update-policy")
    data class UpdatePolicy(val notice: UpdateNotice? = null, val voicePaused: Boolean = false) : Action

    /** A banner or a dialog can be dismissed; a required update cannot. */
    @Serializable @SerialName("dismiss-update")
    data object DismissUpdate : Action

    @Serializable @SerialName("open-app-store")
    data object OpenAppStore : Action

    @Serializable @SerialName("open-support")
    data object OpenSupport : Action

    /** Settings, "Check for updates": on Android updates come from Google Play, so it opens the listing. */
    @Serializable @SerialName("check-for-updates")
    data object CheckForUpdates : Action

    /** Settings, "Privacy policy", and the consent screen's "Learn more" (Google Play: a link inside the app). */
    @Serializable @SerialName("open-privacy-policy")
    data object OpenPrivacyPolicy : Action

    /** The platform's push token arrived (after `turn-on-notifications`): register it with the Mac. */
    @Serializable @SerialName("push-token")
    data class PushToken(val token: String, val previewKey: String? = null) : Action

    /** Scrolled to the top: fetch the next older chunk (contract §5.5; no pagination, chunked loading). */
    @Serializable @SerialName("load-older")
    data object LoadOlder : Action

    /** Photos and files picked or shared, staged and hashed by the platform: into the composer. */
    @Serializable @SerialName("attach")
    data class Attach(val files: List<Attachment>) : Action

    /** Take one picked file back out of the composer. */
    @Serializable @SerialName("remove-attachment")
    data class RemoveAttachment(val id: String) : Action

    /** The + menu's Photos, Camera or Files (`att-menu`): the Mac's limits decide, then the platform's picker. */
    @Serializable @SerialName("pick-attachments")
    data class PickAttachments(val source: AttachSource) : Action

    /** The platform refused an item before staging it (over the Mac's per-file limit while copying). */
    @Serializable @SerialName("attach-refused")
    data class AttachRefused(val name: String, val bytes: Long? = null, val tooLarge: Boolean = true) : Action

    /** The camera permission was refused, or no camera app can take the photo (`att-denied-camera`). */
    @Serializable @SerialName("attach-permission-denied")
    data class AttachPermissionDenied(val source: AttachSource) : Action

    /** "Not now" / "Got it" on a photos-and-files card. */
    @Serializable @SerialName("dismiss-attach-notice")
    data object DismissAttachNotice : Action

    /**
     * Share to Rich from another app (CEO §75): words and/or staged files, sent as one message
     * WITHOUT touching the composer's draft or its pending attachments. [clientId] is chosen by
     * the caller so it can watch that message until the Mac accepts it.
     */
    @Serializable @SerialName("share")
    data class Share(val clientId: String, val text: String = "", val files: List<Attachment> = emptyList()) : Action

    /** Photos and files to Rich (CEO §75), already staged and hashed by the platform. */
    @Serializable @SerialName("send-attachments")
    data class SendAttachments(val files: List<Attachment>, val text: String = "") : Action
}

@Serializable
data class Recording(val id: String, val seconds: Double)

/** A refused action or command. The message is a sentence a person can act on. */
class CoreError(message: String) : IllegalArgumentException(message)

File: richos/mobile/native-ios/Core/Sources/RichOSCore/Action.swift
import Foundation

/// Everything a person — or the platform, or the CLI standing in for either — can do.
///
/// The screens dispatch these and nothing else; `bin/rios headless action '<json>'` dispatches the
/// same values, which is what lets the command line prove a change without a simulator. The JSON
/// spelling is `{"type": "<kebab-case>", …fields}`, the preserved mobile CLI's action envelope
/// (`richos/mobile/cli/mobile.mjs`). Cases arrive with the reducer logic that gives them meaning.
public enum Action: Equatable, Sendable {
    // composer
    case compose(text: String)
    case setAppearance(Appearance)
    // pairing (round-12 group 1)
    case openScanner
    case closeScanner
    /// The QR scanner read this text. It is only ever parsed, never followed.
    case scanned(text: String)
    /// Pasted into the pairing-link sheet.
    case submitPairingLink(text: String)
    /// The OS reports the camera permission (a mirror; nothing parallel is stored).
    case cameraPermission(Permission)
    /// The Mac answered the pairing request (from the transport, or a fixture).
    case pairingAnswered(PairAnswer)
    /// The Mac refused the pairing code (404 after the one retry, contract §2.3).
    case pairingRefused
    /// No Mac answered the pairing request.
    case pairingUnreachable
    /// The Mac answered 200 without `pair-v2`: refused, never fallen back to. The phone has already
    /// signed one "They do not match" so that Mac forgets the key; it now forgets it too.
    case pairingNeedsMacUpdate
    /// "They match" on the six-word check. Pairing v2: the phone then waits for the same press on the Mac.
    case confirmWords
    /// "They do not match" (on the words, or while waiting for the Mac): the Mac forgets this phone
    /// and the phone discards its key (§2.5).
    case rejectWords
    /// What the Mac said about the press ON THE MAC: its answer to the phone's own "They match", or
    /// to one probe of the wait (`MacWait`). `at` is when the answer arrived.
    case macConfirmation(MacConfirmation, at: Int64)
    /// Continue on the consent screen (`pair-consent`).
    case acceptConsent
    case dismissPairingProblem
    /// `pair-blocked`'s "Discard and pair": the unsent messages go (never to whichever Mac is paired
    /// next), then the way to pair opens again.
    case discardUnsentAndPair
    // sheets
    case openSheet(Sheet)
    case closeSheet
    // conversation and outbox (groups 2, 3)
    /// Send the draft. `clientID` and `at` are stamped by the caller (`Action.sendDraftNow()` in the
    /// app, fixed values in a scenario), so the reducer stays pure and a trace replays exactly.
    case sendDraft(clientID: String, at: Int64)
    /// The Mac accepted this message (200, including `duplicate: true`).
    case deliveryAccepted(clientID: String, at: Int64, textSHA256: String? = nil)
    case deliveryFailed(clientID: String, failure: DeliveryFailure, at: Int64)
    /// Time passed (the outbox's retry clock); cheap and never persisted by itself.
    case tick(at: Int64)
    /// "Try now" on the waiting card (`conv-retry`).
    case retryNow(at: Int64)
    /// Remove one of your unsent messages (never one in flight).
    case discardMessage(id: String)
    /// Rows from the live stream or a refresh, merged by id.
    case messagesArrived([Message])
    case replyStarted
    /// The reply's text so far.
    case replyDelta(text: String)
    case replyFinished(Message)
    /// Scrolled to the top: fetch older messages (chunked; never pagination).
    case loadOlder
    case olderLoaded([Message], reachedBeginning: Bool)
    case setFollowing(Bool)
    case rememberReading(ReadingAnchor)
    case setComposerFocus(Bool)
    case openedFromNotification(messageID: String)
    /// A tapped notification for this Mac and conversation, by its reply's reference (SHA-256 hex of
    /// the id). Focuses the reply, fetching older history until it appears (contract §7.3 step 3).
    case openedFromNotificationReference(String)
    /// A share the Share extension saved, taken into the outbox with its own ids and commit bytes.
    case takeShare(SharedIntake, at: Int64)
    case clearFocus
    /// "Hear it" on one of Rich's replies.
    case hearReply(id: String)
    case playbackStarted(id: String)
    case playbackProgress(id: String, progress: Double)
    case playbackEnded
    case stopPlayback
    case dismissToast
    // connection (group 7)
    /// The phone's own network came or went (a path monitor in the app).
    case networkChanged(online: Bool, at: Int64)
    /// The live stream or a request to the Mac failed at the transport level.
    case connectionLost(at: Int64)
    /// The Mac answered again (a stream opened, or a request succeeded).
    case connected(at: Int64)
    /// The transport's probe says why the Mac cannot be reached (evidence-based only).
    case connectionDiagnosed(ConnectionNotice)
    /// What the Mac says it accepts (capability negotiation, contract §12).
    case macCapabilities(text: Bool, voice: Bool)
    /// The Mac answered 403 `{"revoked":true}`: this phone was removed (round-12 `conn-revoked`).
    case pairingRevoked
    /// What the Mac accepts for attachments (from `hello` or the pairing answer); `nil` = none.
    case macAttachmentLimits(AttachmentLimits?)
    /// The Mac registered this phone for reply notifications; its push host id is kept.
    case pushRegistered(hostID: String?)
    /// The app came to the screen (or launched): reconnect and move the outbox.
    case foregrounded(at: Int64)
    /// The app left the screen: the live stream closes; a recording in progress is kept.
    case backgrounded(at: Int64)
    // voice (groups 4, 5, 6)
    /// Touch-down on the microphone. `id` names the recording (`Action.voicePressNow` in the app).
    case voicePress(id: String, width: Double, at: Int64)
    /// VoiceOver's "record hands-free": start a recording already locked.
    case voiceStartLocked(id: String, width: Double, at: Int64)
    /// The OS answered or reports the microphone permission (a mirror; nothing parallel is stored).
    case microphonePermission(Permission)
    /// The microphone-off card's "Not now": the card goes until the next press finds the microphone
    /// off (D03; Android's `dismiss-microphone-card`).
    case dismissMicrophoneCard
    /// The finger moved: offsets from the touch-down point, in points (negative = left / up).
    case voiceMove(dx: Double, dy: Double, at: Int64)
    case voiceRelease(at: Int64)
    /// Tap on the send circle while locked.
    case voiceLockedSend(at: Int64)
    /// Tap on Cancel while locked.
    case voiceLockedCancel(at: Int64)
    /// The system took the touch away (an alert over the app).
    case voiceTouchCanceled(at: Int64)
    case voiceStartFailed(id: String)
    /// The app left the screen or the OS took the audio: the recording is kept, never sent.
    case voiceInterrupted(at: Int64)
    /// The recording's level, 0 to 1, sampled every 100 ms, for the bubble it becomes.
    case voiceLevel(Double)
    /// The end animation finished.
    case voiceSettled
    case sendKept(id: String, at: Int64)
    case discardKept(id: String)
    /// Play one of your own recordings (a kept one or a sent voice bubble).
    case playRecording(id: String)
    // notifications and settings (groups 8, 9)
    case turnOnNotifications
    /// What the OS and the Mac said to a registration (or its absence).
    case notificationsResult(Notifications.Status)
    case turnOffNotifications
    /// "Not now" on the offer card: it is asked once.
    case dismissNotificationOffer
    case setPreviews(Bool)
    /// "Forget this phone" in Settings: opens the confirmation, or the refusal while work is unsent.
    case forgetPairing
    /// The filled, destructive button in the confirmation.
    case confirmForget
    case openSystemSettings
    // update notices (group 10)
    /// The hosted policy's verdict for this app version (contract §5.10).
    case updatePolicy(UpdateNotice?, voicePaused: Bool)
    case dismissUpdate
    case openAppStore
    case openSupport
    /// Settings, "Check for updates": on iPhone updates come from the App Store, so it opens the listing.
    case checkForUpdates
    /// Settings, "Privacy policy" (Apple 5.1.1(i): a link inside the app).
    case openPrivacyPolicy
    // photos and files (round 12.1 groups 12-17)
    case openAttachMenu
    case closeAttachMenu
    /// A row of the + menu: Photos, Camera or Files.
    case pickAttachments(AttachSource)
    /// What the platform staged from Apple's picker, in the order chosen.
    case attachmentsPicked([OutboxFile])
    /// The platform could not take a chosen item (over the Mac's per-file size, or a type it refuses).
    case attachmentRefused(name: String, bytes: Int?, tooLarge: Bool)
    /// The OS refused the camera (or the photo library).
    case attachPermissionDenied(AttachSource)
    /// The × on a tray item.
    case removePendingAttachment(id: String)
    case dismissAttachNotice
}

/// How a delivery attempt ended when it did not succeed (contract §4.2, the reference classification).
public enum DeliveryFailure: Codable, Equatable, Sendable {
    /// No Mac, a fault, or 429: the same bytes again after `afterMs` (the outbox's own clock when nil).
    case retryable(reason: String?, afterMs: Int64? = nil)
    /// A final answer for this item (409, 413, 422, 503 with `retry:false`): it needs the person; the
    /// rest of the queue carries on.
    case refused(reason: String?)
    /// A final answer about the phone, not the message (a final 404, a 403 that is not a revocation):
    /// this item is blocked and the queue stops.
    case refusedStopQueue(reason: String?)
    /// 403 `{"revoked":true}`: this phone was removed from the Mac.
    case revoked
}

extension Action {
    /// The app's touch-down on the microphone: a fresh recording id and the current time.
    public static func voicePressNow(width: Double, clock: any Clock = SystemClock()) -> Action {
        .voicePress(id: UUID().uuidString.lowercased(), width: width, at: clock.nowMs())
    }

    /// The app's way to send the draft: a fresh idempotency key and the current time.
    public static func sendDraftNow(clock: any Clock = SystemClock()) -> Action {
        .sendDraft(clientID: UUID().uuidString.lowercased(), at: clock.nowMs())
    }
}

/// What `POST /api/pair` answers with (contract §2.3), in the core's terms.
public struct PairAnswer: Codable, Equatable, Sendable {
    public var deviceID: String
    /// The Mac's `ca_fingerprint_sha256`, byte for byte as the Mac sent it (the v2 words hash it as
    /// text, so it is never reformatted).
    public var fingerprintHex: String
    public var threadID: String?
    public var macName: String?
    /// This phone's own public key for the pairing: the 65-byte uncompressed point, base64url without
    /// padding. The v2 words hash it; without it no words can be shown, and nothing is paired.
    public var devicePoint: String?
    /// The Mac's `confirm_within_seconds`: how long the person has to press "They match" on the Mac.
    public var confirmWithinSeconds: Double?
    public init(deviceID: String, fingerprintHex: String, threadID: String? = nil, macName: String? = nil,
                devicePoint: String? = nil, confirmWithinSeconds: Double? = nil) {
        self.deviceID = deviceID; self.fingerprintHex = fingerprintHex; self.threadID = threadID; self.macName = macName
        self.devicePoint = devicePoint; self.confirmWithinSeconds = confirmWithinSeconds
    }
}

/// Work a reducer asks for and does not do itself. The reducer stays pure; the effect runner
/// performs these through the ports, so every one can be substituted in a test. An effect with no
/// port wired yet is recorded and skipped, never guessed at.
public enum Effect: Equatable, Sendable {
    /// Write `AppState.persisted` to durable storage.
    case persist
    /// Generate (or reuse) this origin's device key and send `POST /api/pair` with the code.
    case pair(PairLink)
    /// Send the signed six-word answer (contract §2.5). For "They match" the Mac's answer comes back
    /// as `macConfirmation`.
    case confirmFingerprint(matches: Bool)
    /// One probe of the wait for the press on the Mac: a signed read of one backfill row, which the
    /// Mac answers with the awaiting 409 until the press (`MacWait`). Only ever sent on screen.
    case checkMacConfirmation
    /// Discard this phone's key for that origin.
    case forgetIdentity(origin: String)
    /// Open this app's page in iPhone Settings (camera or microphone turned off).
    case openSystemSettings
    /// POST the outbox item's exact bytes (contract §5.2).
    case deliver(clientID: String)
    /// Fetch older history before this message (contract §5.5).
    case loadOlder(before: String?)
    /// Fetch and play the reply's audio (contract §5.6).
    case fetchReplyAudio(messageID: String)
    case stopAudio
    /// Open (or keep) the live connection to the paired Mac.
    case connect
    /// Close it (backgrounded, revoked, forgotten).
    case disconnect
    /// Ask the OS for the microphone (once, on the first deliberate press).
    case requestMicrophone
    case startRecording(id: String)
    /// Stop capturing; `keep` retains the file (sent or kept), otherwise it is removed.
    case stopRecording(id: String, keep: Bool)
    case deleteRecording(id: String)
    /// Remove the phone's copies of attached files (the Mac accepted the message, or it was discarded).
    case deleteAttachments(paths: [String])
    /// The light tick when the lock engages (`UIImpactFeedbackGenerator(.light)`).
    case hapticTick
    case playRecording(id: String)
    /// Ask the OS for notification permission and register with the Mac (native push, contract §7.2).
    case requestNotifications(previews: Bool)
    case unregisterNotifications
    /// Remove every RichOS reply notification still in Notification Center: notifications were turned
    /// off, or the Mac was forgotten (D04).
    case withdrawNotifications
    case openAppStore
    case openSupport
    case openPrivacyPolicy
    /// Present Apple's photo picker, camera or document picker, for at most `maxCount` items.
    case presentPicker(AttachSource, maxCount: Int)
}

/// The one place state changes. Pure: the same state and action always give the same result, so a
/// scenario replays exactly on this Mac and in the simulator.
public enum Reducer {
    public static func reduce(_ state: AppState, _ action: Action) -> (state: AppState, effects: [Effect]) {
        var next = state
        var effects: [Effect] = []
        switch action {
        case .compose(let text):
            next.draft = text
        case .setAppearance(let appearance):
            next.appearance = appearance
        case .openSheet(let sheet):
            next.sheet = sheet
        case .closeSheet:
            next.sheet = nil
        case .openScanner, .closeScanner, .scanned, .submitPairingLink, .cameraPermission, .pairingAnswered,
             .pairingRefused, .pairingUnreachable, .pairingNeedsMacUpdate, .confirmWords, .rejectWords, .macConfirmation,
             .acceptConsent, .dismissPairingProblem, .discardUnsentAndPair:
            PairingReducer.reduce(&next, action, &effects)
        case .networkChanged, .connectionLost, .connected, .connectionDiagnosed, .macCapabilities, .pairingRevoked,
             .macAttachmentLimits, .pushRegistered:
            ConnectionReducer.reduce(&next, action, &effects)
        case .foregrounded, .backgrounded:
            PairingReducer.reduce(&next, action, &effects)
            ConnectionReducer.reduce(&next, action, &effects)
            VoiceReducer.reduce(&next, action, &effects)
            ConversationReducer.reduce(&next, action, &effects)
        case .tick:
            PairingReducer.reduce(&next, action, &effects)
            ConnectionReducer.reduce(&next, action, &effects)
            ConversationReducer.reduce(&next, action, &effects)
            VoiceReducer.reduce(&next, action, &effects)
        case .voicePress, .voiceStartLocked, .microphonePermission, .dismissMicrophoneCard, .voiceMove, .voiceRelease, .voiceLockedSend, .voiceLockedCancel,
             .voiceTouchCanceled, .voiceStartFailed, .voiceInterrupted, .voiceLevel, .voiceSettled, .sendKept, .discardKept, .playRecording:
            VoiceReducer.reduce(&next, action, &effects)
        case .turnOnNotifications, .notificationsResult, .turnOffNotifications, .dismissNotificationOffer, .setPreviews,
             .forgetPairing, .confirmForget, .openSystemSettings:
            SettingsReducer.reduce(&next, action, &effects)
        case .updatePolicy, .dismissUpdate, .openAppStore, .openSupport, .checkForUpdates, .openPrivacyPolicy:
            UpdateReducer.reduce(&next, action, &effects)
        case .openAttachMenu, .closeAttachMenu, .pickAttachments, .attachmentsPicked, .attachmentRefused, .attachPermissionDenied,
             .removePendingAttachment, .dismissAttachNotice:
            ConversationReducer.reduceAttachments(&next, action, &effects)
        default:
            ConversationReducer.reduce(&next, action, &effects)
        }
        // Persist only when something durable changed: a touch-rate action never writes a file.
        if next.persisted != state.persisted { effects.insert(.persist, at: 0) }
        return (next, effects)
    }
}
